Test results to Slack: add support for notifications rules (#25830)
Automattic/jetpack: projects/github-actions/test-results-to-slack/README.md

- (Required) `github_token` is a GitHub Access Token used to access GitHub's API. The token should be stored in a [secret](https://docs.github.com/en/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository)
- (Required) `slack_token` is the Auth token of a bot that is installed on your Slack workspace. The token should be stored in a [secret](https://docs.github.com/en/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository). See the [instructions here](https://slack.com/intl/en-hu/help/articles/115005265703-Create-a-bot-for-your-workspace) on how to create a bot.

> - (Required) `slack_channel` is the Slack channel ID where the messages will be sent to. See below how to find the channel ID. Check the channel's details in you Slack app to get find the channel ID.

Reviewer: ```suggestion
- (Required) `slack_channel` is the Slack channel ID where the messages will be sent to. See below how to find the channel ID. Check the channel's details in your Slack app to find the channel ID.
```

Author: Done and done with [0ab3521](https://github.com/Automattic/jetpack/pull/25830/commits/0ab352190a6f86b2ddf795f37064f0ebecb5ce70)

Author: And with [1f5660a](https://github.com/Automattic/jetpack/pull/25830/commits/1f5660a69278bec6df1ae7f642e12b1c1efae989). Apparently I need 2 commits to fix a typo.

- (Optional) `slack_icon_emoji` is the icon emoji to use for messages. If not set it will use your app's default icon.
- (Optional) `suite_name` is the name of the test suite. It will be included in the message, and it can also be used to define notification rules. See more in the Rules section.
- (Optional) `rules_configuration_path` is the path to the configuration file that defines the rules. See more in the Rules section.

### Rules

You can configure different rules, to send notifications in multiple channels.

There are two types of rules: refs rules, used to send notifications for specific branches or tags and suite rules, used to send notifications for specific test suites.

#### Refs rules

You can create as many rules as you want. For each rule, you need to define the ref type and ref name to match, and a list of channels to send the notification in case of a match. Optionally you can also define whether to exclude the default channel. By default, the default channel is not excluded and a notification will also be sent there.

Example:

```json
{
  "refs": [
	{

> 	  "type": "branch",

Reviewer: What other ref types are available? Do you think it would make sense to mention them with a short description?

Author: I expect that if one is setting this up, there's already another build or tests workflow running for that refType, so they should already know what they have and need. 
Nevertheless, it wouldn't hurt mentioning something, so I added a small example with [8ad7721](https://github.com/Automattic/jetpack/pull/25830/commits/8ad772199a2a258f759be5fe882ee7dff732e23e)

	  "name": "trunk",
	  "channels": [
		"CHANNEL_ID_1"
	  ],
	  "excludeDefaultChannel": true
	},
	{
	  "type": "branch",
	  "name": "dev",
	  "channels": [
		"CHANNEL_ID_2"
	  ]
	}
  ]
}
```

In the example, for runs on branch trunk, a notification will be sent to `CHANNEL_ID_1`. For runs on branch dev, a notification will be sent to `CHANNEL_ID_2` and to the default channel.

#### Suites rules

You can create as many suites rules as you want. For each rule, you need to define the suite name to match, and a list of channels to send the notification in case of a match. Optionally you and also define whether to exclude the default channel. By default, the default channel is not excluded.

Example:

```json
{
  "suites": [
	{
	  "name": "Smoke tests",
	  "channels": [
		"CHANNEL_ID_1"
	  ]
	}
  ]
}
```

> In the example, for runs with `suite_name` set to "Smoke tests", a notification will be sent to `CHANNEL_ID_1` and to the default channel.

Reviewer: Should we mention/link to how to obtain channel IDs?

Author: Good point. I couldn't however find an article about that in Slack's docs. I don't want to add exact steps in our docs, it would add a maintainability burden, but I think I found a good middle ground in [8ad7721](https://github.com/Automattic/jetpack/pull/25830/commits/8ad772199a2a258f759be5fe882ee7dff732e23e)
